app/Main.hs

```diff
@@ -3,40 +3,95 @@
+import Middleware (forceTls, prettifyError)
 
 main :: IO ()
--- main = getEnv "PORT" >>= someFunc . read
 main = do
-  ms <- settingsFromEnv UUID.toText UUID.toText
-  store <- getDatabaseConfig >>= newDBStore
-  o <- sGetNotificationChan store
-  let actor = newActor "NaCl" getCurrentTime
-  go ms store actor o
+    -- Read a bunch of environment config:
+    port <- read <$> getEnv "PORT"
+    allowInsecure <- maybe False (not . null) <$> lookupEnv "ALLOW_INSECURE"
+    domain <- Text.pack <$> getEnv "DOMAIN"
+    salt <- Text.pack <$> getEnv "REGISTRATION_UUID_SALT"
+    -- The username and password that we will come in with when querying our API
+    -- for a list of all registered email addresses:
+    username <- UTF8.fromString <$> getEnv "REGISTRATION_USERNAME"
+    password <- UTF8.fromString <$> getEnv "REGISTRATION_PASSWORD"
+    ms <- mailerSettingsFromEnv (formatVLink domain) (formatULink domain)
+    dbConfig <- getDatabaseConfig
+
+    -- Do a bunch of initialisation:
+    store <- newDBStore dbConfig
+    o <- sGetNotificationChan store
+    let actor = newActor salt getCurrentTime
+
+    let authMiddlware = buildAuth username password
+    let icp = interestedCollectionPost actor store
+    let ir = interestedResource actor store
+    let app = prettifyError $ dispatch $ root icp ir authMiddlware
+
+    withAsync (mailer ms store (U.readChan o)) $ \a -> do
+        link a
+        if allowInsecure
+        then Warp.run port app
+        else Warp.run port $ forceTls app
+
+
+formatVLink :: Text -> UUID -> Text
+formatVLink = formatActionLink "verify"
+
+
+formatULink :: Text -> UUID -> Text
+formatULink = formatActionLink "unsubscribe"
+
+
+formatActionLink :: Text -> Text -> UUID -> Text
+formatActionLink verb domain uuid =
+    "https://" <> domain <> "/interested/" <> UUID.toText uuid
+    <> "?action=" <> verb
+
+
+buildAuth :: BS.ByteString -> BS.ByteString -> Wai.Middleware
+buildAuth username password = basicAuth isAllowed "Concert API"
   where
-    parseUuidThen f uuid = maybe (putStrLn "rubbish uuid") f $ UUID.fromString uuid
-    go ms store actor o =
-        withAsync (mailer ms store (U.readChan o)) $
-          \_ -> do
-            putStrLn "Command pls: s <email>, v <uuid>, u <uuid>, g <uuid>, q"
-            input <- getLine
-            case input of
-              's':' ':email -> let e = Text.pack email in
-                  aSubmitEmailAddress actor store e >> go ms store actor o
-              'v':' ':uuid ->
-                  parseUuidThen (\u -> aVerify actor store u) uuid >>
-                  go ms store actor o
-              'u':' ':uuid ->
-                  parseUuidThen (\u -> aUnsubscribe actor store u) uuid >>
-                  go ms store actor o
-              'g':' ':uuid -> parseUuidThen (getAndShowState store) uuid >>
-                  go ms store actor o
-              'q':_ -> sSendShutdown store
-              _ -> putStrLn "Narp, try again" >> go ms store actor o
+    isAllowed u p
```

Comment:
This isn't a syntax I'm used to, feel like it should only have to say pure once somehow.

Reply:
The `|` is syntax for a guard, so you can define many equations with different predicates tried in order. I think you're right, I could have used `(pure . isAllowed)` on line 73.